File: go/refactoring/golden_master_test.go
package refactoring

import (
	"testing"
	"time"

	"github.com/bradleyjkemp/cupaloy/v2"
)

var counter = 0

// ReceiptPrinterWithoutDiscountForTest is a test version with controlled date and no discount
type ReceiptPrinterWithoutDiscountForTest struct {
	*ReceiptPrinter
}

func (r *ReceiptPrinterWithoutDiscountForTest) GetCurrentDate() time.Time {
	return time.Date(2022, 2, 1, 0, 0, 0, 0, time.UTC)
}

func (r *ReceiptPrinterWithoutDiscountForTest) discount() float64 {
	return 0.0
}

// ReceiptPrinterWithDiscountForTest is a test version with controlled date and fixed discount
type ReceiptPrinterWithDiscountForTest struct {
	*ReceiptPrinter
}

func (r *ReceiptPrinterWithDiscountForTest) GetCurrentDate() time.Time {
	return time.Date(2022, 2, 1, 0, 0, 0, 0, time.UTC)
}

func (r *ReceiptPrinterWithDiscountForTest) discount() float64 {
	return 0.05
}

func TestShouldPrintAReceipt(t *testing.T) {
	// Given a customer
	customer := "Ana"
	// Given a number of items
	item := 1
	// Given quantity
	quantity := 1

	counter++
	order := GenerateOrder("ORD-1", customer, item, quantity)
	printer := &ReceiptPrinterWithoutDiscountForTest{
		ReceiptPrinter: NewReceiptPrinter(),
	}
	receipt := printer.Print(order)

	cupaloy.SnapshotT(t, receipt)
}

File: php/tests/GoldenMasterTest.php
<?php

namespace Tests;

use PHPUnit\Framework\TestCase;
use Refactoring\GoldenMaster\ReceiptPrinter;
use function Refactoring\GoldenMaster\generateOrder;

class ReceiptPrinterWithoutDiscountForTest extends ReceiptPrinter
{
    protected function getCurrentDate(): \DateTime
    {
        return new \DateTime('2022-02-01');
    }

    protected function discount(): float
    {
        return 0.0;
    }
}

class ReceiptPrinterWithDiscountForTest extends ReceiptPrinter
{
    protected function getCurrentDate(): \DateTime
    {
        return new \DateTime('2022-02-01');
    }

    protected function discount(): float
    {
        return 0.05;
    }
}

class GoldenMasterTest extends TestCase
{
    private static int $counter = 0;

    public function testShouldPrintAReceipt(): void
    {
        // Given a customer
        $customer = 'Ana';
        // Given a number of items
        $item = 1;
        // Given quantity
        $quantity = 1;

        self::$counter++;
        $order = generateOrder('ORD-' . self::$counter, $customer, $item, $quantity);
        $receipt = (new ReceiptPrinterWithoutDiscountForTest())->print($order);

        $this->assertMatchesSnapshot($receipt);
    }
}
